go/cmd/dolt/commands/sqlserver/server.go

```diff
@@ -63,335 +64,470 @@ func Serve(
-			startError = err
-			return
-		}
-		clusterController.RegisterGrpcServices(sqlEngine.NewDefaultContext, clusterRemoteSrv.GrpcServer())
+			args, err := clusterController.RemoteSrvServerArgs(sqlEngine.NewDefaultContext, remotesrv.ServerArgs{
+				Logger: logrus.NewEntry(lgr),
+			})
+			if err != nil {
+				lgr.Errorf("error creating SQL engine context for remotesapi server: %v", err)
+				return err
+			}
 
-		listeners, err := clusterRemoteSrv.Listeners()
-		if err != nil {
-			lgr.Errorf("error starting remotesapi server listeners for cluster config on %s: %v", clusterController.RemoteSrvListenAddr(), err)
-			startError = err
-			return
-		}
+			clusterRemoteSrvTLSConfig, err := LoadClusterTLSConfig(serverConfig.ClusterConfig())
+			if err != nil {
+				lgr.Errorf("error starting remotesapi server for cluster config, could not load tls config: %v", err)
+				return err
+			}
+			args.TLSConfig = clusterRemoteSrvTLSConfig
 
-		go clusterRemoteSrv.Serve(listeners)
-		go clusterController.Run()
+			clusterRemoteSrv, err = remotesrv.NewServer(args)
+			if err != nil {
+				lgr.Errorf("error creating remotesapi server on port %d: %v", *serverConfig.RemotesapiPort(), err)
+				return err
+			}
+			clusterController.RegisterGrpcServices(sqlEngine.NewDefaultContext, clusterRemoteSrv.GrpcServer())
 
-		clusterController.ManageQueryConnections(
-			mySQLServer.SessionManager().Iter,
-			sqlEngine.GetUnderlyingEngine().ProcessList.Kill,
-			mySQLServer.SessionManager().KillConnection,
-		)
+			clusterRemoteSrvListeners, err = clusterRemoteSrv.Listeners()
+			if err != nil {
+				lgr.Errorf("error starting remotesapi server listeners for cluster config on %s: %v", clusterController.RemoteSrvListenAddr(), err)
+				return err
+			}
+			return nil
+		},
+		Run: func(context.Context) {
+			if clusterRemoteSrv == nil {
+				return
+			}
+			clusterRemoteSrv.Serve(clusterRemoteSrvListeners)
+		},
+		Stop: func() error {
+			if clusterRemoteSrv == nil {
+				return nil
+			}
+			clusterRemoteSrv.GracefulStop()
+			return nil
+		},
 	}
+	controller.Register(RunClusterRemoteSrv)
 
-	if ok, f := mrEnv.IsLocked(); ok {
-		startError = env.ErrActiveServerLock.New(f)
-		return
+	RunClusterController := &svcs.Service{
+		Init: func(context.Context) error {
+			if clusterController == nil {
+				return nil
+			}
+			clusterController.ManageQueryConnections(
+				mySQLServer.SessionManager().Iter,
+				sqlEngine.GetUnderlyingEngine().ProcessList.Kill,
+				mySQLServer.SessionManager().KillConnection,
+			)
+			return nil
+		},
+		Run: func(context.Context) {
+			if clusterController == nil {
+				return
+			}
+			clusterController.Run()
+		},
+		Stop: func() error {
+			if clusterController == nil {
+				return nil
+			}
+			clusterController.GracefulStop()
+			return nil
+		},
 	}
+	controller.Register(RunClusterController)
 
-	if err = mrEnv.Lock(serverLock); err != nil {
-		startError = err
-		return
+	RunSQLServer := &svcs.Service{
```

Comment:
This is broken up from InitSQLServer, but it's possible they should be together. We need access to the mySQLServer instance for the cluster controller, but maybe there are other ways to handle that.

AFAICT, InitSQLServer opens the listeners, and so it should be paired up with the Close at least. But we want the Close() of these listeners to be the first thing we do on shutdown, before we shutdown the cluster controller for example.

We could add some more code to wrangle things further given the current setup – for example, we could set a flag or move the pointer to a new variable so that InitSQLServer is responsible for closing these things until we get far enough in the boot process, and after that point RunSQLServer becomes responsible.